### ISSUE TITLE
[Potential Reuse of Nonces in AES-GCM]

### ISSUE THREAD

**Issue by cryptopathe** [ANSSI-FR/MLA#12]
When operating AES-GCM, nonces must never be reused under the same key, to avoid a complete collapse of confidentiality and authenticity.

Looking at the archive format, it looks like that a global 64-bit nonce `nonce` is generated securely at random, and for each `DataBlock` `i`, with `i = 0, 1, ...`, a 96-bit nonce is built by concatenating `nonce` with the 32-bit big-endian representation of `i`. Each `DataBlock` has potentially 128*1024 bytes. Internally, the  96-bit nonce is going to be incremented by GCM for every 16-byte block, which means that the internal counter values for `DataBlock` `0` will be `nonce||uint32(0)`, `nonce||uint32(1)`, ..., `nonce||uint32(16383)`, the internal counter values for `DataBlock` `1` will be `nonce||uint32(1)`, ..., `nonce||uint32(16384)`, etc.

In a nutshell, the sequences of counters will be mostly overlapping in all `DataBlock`. A possibility to correct this security flaw consists in generating the nonce at random for each `DataBlock`, or to derive it with `HKDF`.

**Comment by cryptopathe** [comment]
Nope, I was wrong, sorry about that. The GCM standard ensure that the 128-bit counters does not overlap internally.

**Comment by commial** [comment]
Still, one more look is always welcomed. Better be safe than sorry...
